src/Gui/Inventor/Draggers/Gizmo.cpp

```diff
@@ -0,0 +1,697 @@
+
+    GizmoPlacement placement = gizmo->getDraggerPlacement();
+
+    draggerContainer->translation = Base::convertTo<SbVec3f>(placement.pos);
+    draggerContainer->setPointerDirection(placement.dir);
+
+    translationSensor.setData(this);
+    translationSensor.setFunction(translationSensorCB);
+    translationSensor.setPriority(0);
+    SoSFVec3f& translation = gizmo->getDraggerContainer()->getDragger()->translation;
+    translationSensor.attach(&translation);
+    translation.touch();
+
+    automaticOrientation = true;
+}
+
+void RotationGizmo::translationSensorCB(void* data, SoSensor* sensor)
+{
+    assert(data);
+    auto sudoThis = static_cast<RotationGizmo*>(data);
+    assert(sensor);
+    auto translationSensor = static_cast<SoFieldSensor*>(sensor);
+
+    GizmoPlacement placement = sudoThis->linearGizmo->getDraggerPlacement();
+
+    SbVec3f translation = static_cast<SoSFVec3f*>(translationSensor->getAttachedField())->getValue();
+    float yComp = translation.getValue()[1];
+    SbVec3f dir = placement.dir;
+    dir.normalize();
+    sudoThis->draggerContainer->translation = placement.pos + dir * (yComp + sudoThis->sepDistance);
+}
+
+void RotationGizmo::placeBelowLinearGizmo(LinearGizmo* gizmo)
+{
+    linearGizmo = gizmo;
+
+    GizmoPlacement placement = gizmo->getDraggerPlacement();
+
+    draggerContainer->translation = Base::convertTo<SbVec3f>(placement.pos);
+    draggerContainer->setPointerDirection(-placement.dir);
+
+    translationSensor.setData(this);
+    translationSensor.setFunction(translationSensorCB);
+    translationSensor.setPriority(0);
+    SoSFVec3f& translation = gizmo->getDraggerContainer()->getDragger()->translation;
+    translationSensor.attach(&translation);
+    translation.touch();
+}
+
+double RotationGizmo::getRotAngle()
+{
+    double rotAngle = dragger->rotationIncrementCount.getValue()
+        * dragger->rotationIncrement.getValue();
+
+    return (rotAngle - addFactor) / multFactor;
+}
+
+void RotationGizmo::setRotAngle(double angle)
+{
+    angle = multFactor * angle + addFactor;
+    dragger->rotation = SbRotation({0, 0, 1.0f}, static_cast<float>(angle));
+}
+
+void RotationGizmo::setGeometryScale(float scale)
+{
+    dragger->geometryScale = SbVec3f(scale, scale, scale);
+}
+
+SoRotationDraggerContainer* RotationGizmo::getDraggerContainer()
+{
+    assert(draggerContainer);
+    return draggerContainer;
+}
+
+void RotationGizmo::draggingStarted()
+{
+    initialValue = property->value().getValue();
+    dragger->rotationIncrementCount.setValue(0);
+
+    if (delayedUpdateEnabled()) {
+        property->blockSignals(true);
+    }
+}
+
+void RotationGizmo::draggingFinished()
+{
+    if (delayedUpdateEnabled()) {
+        property->blockSignals(false);
+        property->valueChanged(property->value().getValue());
+    }
+}
+
+double normalizeIn360(double value)
+{
+    value = std::fmod(value, 360.0);
+    if (value < 0.0) {
+        value += 360.0;
+    }
+
+    return value;
+}
```

Comment:
Base::fmod does the same thing, you don't need that function

Reply:
`std::fmod` yes, but not `Base::fmod` - it implements the same logic as you have here:

https://github.com/FreeCAD/FreeCAD/blob/5a0c03c891ed7e840d4c5833f959a16cabca4fd4/src/Base/Tools.h#L176-L181

Reply:
Oh, I misread your message. I will update it